### PR TITLE
PHPLIB-1001, PHPLIB-1010, PHPLIB-1011: Improvements to $$type operator

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPLIB-1001
https://jira.mongodb.org/browse/PHPLIB-1010
https://jira.mongodb.org/browse/PHPLIB-1011

Also revises instantiation of deprecated BSON types and fixes an assertion message.

### CODE REVIEW

File: tests/UnifiedSpecTests/Constraint/IsBsonType.php [hunk 1]
```diff
@@ -167,6 +168,9 @@ private function doMatches($other): bool
             case 'maxKey':
                 return $other instanceof MaxKeyInterface;
 
+            case 'number':
+                return is_int($other) || (PHP_INT_SIZE == 4 && $other instanceof Int64) || is_float($other) || $other instanceof Decimal128Interface;
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/Constraint/IsBsonType.php:172]: Out of curiosity, is there a reason for limiting the `Int64` check to 32-bit platforms? Technically we won't have an `Int64` instance on 64-bit platforms, but is there a practical reason for the check?

File: tests/UnifiedSpecTests/Constraint/IsBsonTypeTest.php [hunk 2]
```diff
@@ -38,9 +38,10 @@ public function testConstraint($type, $value): void
 
     public function provideTypes()
     {
-        $undefined = toPHP(fromJSON('{ "undefined": {"$undefined": true} }'));
-        $symbol = toPHP(fromJSON('{ "symbol": {"$symbol": "test"} }'));
-        $dbPointer = toPHP(fromJSON('{ "dbPointer": {"$dbPointer": {"$ref": "phongo.test", "$id" : { "$oid" : "5a2e78accd485d55b405ac12" }  }} }'));
+        $undefined = toPHP(fromJSON('{ "x": {"$undefined": true} }'))->x;
+        $symbol = toPHP(fromJSON('{ "x": {"$symbol": "test"} }'))->x;
+        $dbPointer = toPHP(fromJSON('{ "x": {"$dbPointer": {"$ref": "db.coll", "$id" : { "$oid" : "5a2e78accd485d55b405ac12" }  }} }'))->x;
+        $long = PHP_INT_SIZE == 4 ? unserialize('C:18:"MongoDB\BSON\Int64":38:{a:1:{s:7:"integer";s:10:"4294967296";}}') : 4294967296;
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/Constraint/IsBsonTypeTest.php:44]: Our tests make the best case for providing a constructor for `Int64`. Can't wait to eventually get rid of it.

File: tests/UnifiedSpecTests/Constraint/IsBsonTypeTest.php [hunk 3]
```diff
@@ -89,10 +94,20 @@ public function testAnyOf(): void
         $c = IsBsonType::anyOf('double', 'int');
 
         $this->assertResult(true, $c, 1, 'int is double or int');
-        $this->assertResult(true, $c, 1.4, 'int is double or int');
+        $this->assertResult(true, $c, 1.4, 'double is double or int');
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/Constraint/IsBsonTypeTest.php:97]: Good catch!

**Review by alcaeus — APPROVED** [review]

File: tests/UnifiedSpecTests/Constraint/IsBsonType.php (hunk 1 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/IsBsonType.php:172]: This likely goes back to the legacy test format's [DocumentMatchConstraint](https://github.com/mongodb/mongo-php-library/blob/master/tests/SpecTests/DocumentsMatchConstraint.php), from which we derived IsBsonType.

I'll note that the corresponding [Matches constraint prepares 64-bit integers](https://github.com/mongodb/mongo-php-library/blob/ecdacff94528eb4bd27c5a053b34bbb2d457eb40/tests/UnifiedSpecTests/Constraint/Matches.php#L427) such that we should never have an Int64 type on a 64-bit platform; however, that doesn't mean IsBsonType needs to be this strict. This PR seems like a good opportunity to relax that logic, so I'll make the change.

Side note: PHPLIB-1001 only applies the the unified test runner, so DocumentMatchConstraint will continue to not support "number" in its `$$type` operator implementation.

File: tests/UnifiedSpecTests/Constraint/IsBsonTypeTest.php (hunk 2 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/IsBsonTypeTest.php:44]: Remember that a constructor would provide users a foot-gun allowing them to force 32-bit values to serialize as 64-bit, only to have BSON decoding convert them back to basic integers. Previous discussions on this:

 * https://github.com/mongodb/mongo-php-library/issues/589
 * https://github.com/mongodb/mongo-php-driver/issues/525

Unless we implement some BSON codec API that allows users to coerce integer serialization (e.g. always use 64-bit BSON types), I think the current behavior is less likely to confuse folks while noting we can't please everyone.

**Review by jmikola — COMMENTED** [review]
After I implemented [PHPLIB-1010](https://jira.mongodb.org/browse/PHPLIB-1010) to relax the Int64 logic I realized that the legacy `$$type` operator (a) did not have a default case for unsupported values and (b) never added support for an array of multiple types. There was also a separate bug where we used "boolean" instead of "bool" (actual type from the server operator).

The assertion method also had a `string` type hint (possibly added recently), so I imagine we'd have encountered an error attempting to cast an array to a string if we ever encountered a test using multiple types.

Anyway, this seemed like a good time to just refactor the entire legacy constraint to use IsBsonType, which added the necessary functionality and deleted a fair bit of code in the process (including some legacy code paths for PHPUnit < 6.5).

File: tests/UnifiedSpecTests/Constraint/IsBsonTypeTest.php (hunk 2 above)
Comment (alcaeus) [reply tests/UnifiedSpecTests/Constraint/IsBsonTypeTest.php:45]: I wasn't suggesting we add a constructor - apologies if this was misunderstood. I was alluding to getting rid of Int64 if and when we drop support for 32-bit platforms. It's wishful thinking at this point, but I hope that x86 goes out of style sometime this century ^^

**Review by alcaeus — APPROVED** [review]
LGTM

**Pull request merged**